crates/sui-sdk/examples/utils.rs

```diff
@@ -0,0 +1,305 @@
+    let addresses = addresses
+        .into_iter()
+        .filter(|address| address != &active_address)
+        .collect::<Vec<_>>();
+    let recipient = addresses
+        .get(0)
+        .expect("Cannot get the recipient address needed for writing operations. Aborting");
+
+    Ok((client, active_address, *recipient))
+}
+
+/// Return a sui client to interact with the APIs and an active address from the local wallet.
+///
+/// This function sets up a wallet in case there is no wallet locally,
+/// and ensures that the active address of the wallet has SUI on it.
+/// If there is no SUI owned by the active address, then it will request
+/// SUI from the faucet.
+pub async fn setup_for_read() -> Result<(SuiClient, SuiAddress), anyhow::Error> {
+    let client = SuiClientBuilder::default().build_testnet().await?;
+    println!("Sui testnet version is: {}", client.api_version());
+    let mut wallet = retrieve_wallet().await?;
+    assert!(wallet.get_addresses().len() >= 2);
+    let active_address = wallet.active_address()?;
+
+    println!("Local wallet active address is: {active_address}");
+
+    // make sure we have some SUI (5_000_000 MIST) on this address
+    let coin = fetch_coin(&client, &active_address).await?;
+
+    if coin.is_none() {
+        request_tokens_from_faucet(active_address).await?;
+    }
+
+    Ok((client, active_address))
+}
+
+/// Request tokens from the Faucet for the given address
+pub async fn request_tokens_from_faucet(address: SuiAddress) -> Result<(), anyhow::Error> {
+    let address_str = address.to_string();
+    let json_body = json![{
+        "FixedAmountRequest": {
+            "recipient": &address_str
+        }
+    }];
+
+    // make the request to the faucet JSON RPC API for coin
+    let client = Client::new();
+    let resp = client
+        .post(SUI_FAUCET)
+        .header("Content-Type", "application/json")
+        .json(&json_body)
+        .send()
+        .await?;
+    println!(
+        "Faucet request for address {address_str} has status: {}",
+        resp.status()
+    );
+    println!("Waiting for the faucet to complete the gas request ...");
+    let faucet_resp: FaucetResponse = resp.json().await?;
+    let task_id = if faucet_resp.error.is_none() {
+        faucet_resp.task
+    } else {
+        return Err(anyhow!(
+            "Faucet request was unsuccessful. Error is {:?}",
+            faucet_resp.error
+        ));
+    };
+
+    println!("Faucet request task id: {task_id}");
+
+    let json_body = json![{
+        "GetBatchSendStatusRequest": {
+            "task_id": &task_id
+        }
+    }];
+
+    loop {
+        let resp = client
+            .get("https://faucet.testnet.sui.io/v1/status")
+            .header("Content-Type", "application/json")
+            .json(&json_body)
+            .send()
+            .await?;
+        let text = resp.text().await?;
+        if text.contains("SUCCEEDED") {
+            std::thread::sleep(Duration::from_secs(5));
```

Comment:
Why sleep even in the case when it's succeeded?  Is it to make sure the fullnode sees the change as well? If so, you should just query the fulnode for the owner of one of the coins returned, to make sure it's your address (in a separate loop).

Reply:
Yes, sometimes the node does not seem to have gotten the information, so sleeping was the hacky way to ensure that when calling whatever apis we can get the data we expect.

Reply:
Just to give you an example, two runs, exact same code. 
```
**1st run**
cargo run --example coin_read_api
sui/target/debug/examples/coin_read_api`
Sui testnet version is: 1.7.1
Local wallet active address is: 0x86970aaffac0bbaaae52364b5d5fca2b63614edef3199ba8544dffce2a5e972c
Faucet request for address 0x86970aaffac0bbaaae52364b5d5fca2b63614edef3199ba8544dffce2a5e972c has status: 202 Accepted
Waiting for the faucet to complete the gas request ...
Faucet request task id: 04650412-6116-41bc-a26d-b2e24f7ea030
 *** Coins ***
Page { data: [Coin { coin_type: "0x2::sui::SUI", coin_object_id: 0x054a4b5b724792fb6e45e878d81c24f4a5c98109c1abe7c6aa94da6899aeb6ce, version: SequenceNumber(965506), digest: o#46DZApzPRNBT1iVyD1sZWohHmQP1oae9vTCbRK
 *** Coins ***
[...] 

Second run, after removing the ~/.sui folder

rm -rf ~/.sui
cargo run --example coin_read_api
Running `/sui/target/debug/examples/coin_read_api`
Sui testnet version is: 1.7.1
Local wallet active address is: 0x7664be11b0bf96cb33350c416d75154d3f3e4866fa081b2ed138acbb3cc8841d
Faucet request for address 0x7664be11b0bf96cb33350c416d75154d3f3e4866fa081b2ed138acbb3cc8841d has status: 202 Accepted
Waiting for the faucet to complete the gas request ...
Faucet request task id: aa5aa0b2-7657-4650-9258-592e83275fbc
 *** Coins ***
Page { data: [], next_cursor: None, has_next_page: false }
 *** Coins ***

 *** All coins ***
Page { data: [Coin { coin_type: "0x2::sui::SUI", coin_object_id: 0xe5d9779b62edef532403aa4ab48a0444055bfc1e7a7c7cd6fea55382a94bcaa5, version: SequenceNumber(701378), digest: o#HdrGxB3XUrPAAWx5XWpVB1aEP3RDMWXPjw7Qss
 *** All coins ***
```